### PR TITLE
Properly handle H2 upgrade with data packets

### PR DESCRIPTION
**Pull request by isaacrivriv**
Added logic to handle misparsed data on HTTP2 upgrade
- Detect if data is missing when Upgrade comes
- Queue up data locally before handling upgrade
- Reset everything after

### CODE REVIEW

**Comment by isaacrivriv** [comment]
#build

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/inbound/HttpInputStreamImpl.java [hunk 1]
```diff
@@ -468,6 +481,27 @@ public void restart() {
         }
     }
 
+    public void resetReaderChannelMultiRead() {
```
Comment (pmd1nh) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/inbound/HttpInputStreamImpl.java]: May be change the method name to "setReadFromChannelComplete" ?

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java [hunk 2]
```diff
@@ -1338,6 +1339,40 @@ public boolean handleHTTP2UpgradeRequest(Map<String, String> http2Settings) {
...
+                    }
+                    for (int n; (n = body.read(inBytes)) != -1;) {}
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "Finished request read loop.");
+                    }
+                }catch(Exception e){
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "Got exception reading request and queueing up data. Can't handle request upgrade to HTTP2.", e);
+                    }
+                    vc.getStateMap().put(h2InitError, true);
```
Comment (pmd1nh) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java:1363]: if there is an exception, how is the channel multi read buffer postDataBuffer get cleanup?  It depends on the dataAlreadyReadFromChannel (true) (in the non custom property case) which now can't be set due to an exception.

**Comment by isaacrivriv** [comment]
#build

**Comment by isaacrivriv** [comment]
#libby

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java [hunk 3]
```diff
@@ -310,15 +310,15 @@ final public void resetMsgParsedState() {
      *
      * @return boolean
      */
-    final protected boolean isBodyComplete() {
+    final public boolean isBodyComplete() {
```
Comment (pmd1nh) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java:313]: Update the Copy right year for this class.

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/inbound/HttpInputStreamImpl.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2009 IBM Corporation and others.
+ * Copyright (c) 2009, 2023 IBM Corporation and others.
```
Comment (pmd1nh) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/inbound/HttpInputStreamImpl.java]: Copyright 2009, 2024

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java [hunk 5]
```diff
@@ -35,6 +35,7 @@
 import com.ibm.ws.http.channel.internal.inbound.HttpInboundChannel;
 import com.ibm.ws.http.channel.internal.inbound.HttpInboundLink;
 import com.ibm.ws.http.channel.internal.inbound.HttpInboundServiceContextImpl;
+import com.ibm.ws.http.channel.internal.inbound.HttpInputStreamImpl;
```
Comment (pmd1nh) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java:38]: update Copyright

**Comment by isaacrivriv** [comment]
Build hasn't started so will work with the same build link

**Comment by isaacrivriv** [comment]
#libby

**Comment by isaacrivriv** [comment]
#build

**Review by pmd1nh — APPROVED** [review]

**Comment by isaacrivriv** [comment]
#libby

**Pull request merged**